xds/src/test/java/io/grpc/xds/XdsClientImplTest.java

```diff
@@ -1080,6 +1098,588 @@ public void routeConfigurationRemovedNotifiedToWatcher() {
+            ImmutableList.of(
+                buildLocalityLbEndpoints("region1", "zone1", "subzone1",
+                    ImmutableList.of(
+                        buildLbEndpoint("192.168.0.1", 8080, HealthStatus.HEALTHY, 2),
+                        buildLbEndpoint("192.132.53.5", 80, HealthStatus.UNHEALTHY, 5)),
+                    1, 0)),
+            ImmutableList.<Policy.DropOverload>of())));
+
+    DiscoveryResponse response =
+        buildDiscoveryResponse("0", clusterLoadAssignments,
+            XdsClientImpl.ADS_TYPE_URL_EDS, "0000");
+    responseObserver.onNext(response);
+
+    // Client sent an ACK EDS request.
+    verify(requestObserver)
+        .onNext(eq(buildDiscoveryRequest("0", "cluster-foo.googleapis.com",
+            XdsClientImpl.ADS_TYPE_URL_EDS, "0000")));
+
+    ArgumentCaptor<EndpointUpdate> endpointUpdateCaptor1 = ArgumentCaptor.forClass(null);
+    verify(watcher1).onEndpointChanged(endpointUpdateCaptor1.capture());
+    EndpointUpdate endpointUpdate1 = endpointUpdateCaptor1.getValue();
+    assertThat(endpointUpdate1.getClusterName()).isEqualTo("cluster-foo.googleapis.com");
+    assertThat(endpointUpdate1.getLocalityLbEndpointsMap())
+        .containsExactly(
+            new Locality("region1", "zone1", "subzone1"),
+            new LocalityLbEndpoints(
+                ImmutableList.of(
+                    new LbEndpoint("192.168.0.1", 8080, 2, true),
+                    new LbEndpoint("192.132.53.5", 80,5, false)),
+                1, 0));
+
+    // Another endpoint watcher interested in the same cluster is added.
+    EndpointWatcher watcher2 = mock(EndpointWatcher.class);
+    xdsClient.watchEndpointData("cluster-foo.googleapis.com", watcher2);
+
+    // Since the client has received endpoint update for this cluster before, cached result is
+    // notified to the newly added watcher immediately.
+    ArgumentCaptor<EndpointUpdate> endpointUpdateCaptor2 = ArgumentCaptor.forClass(null);
+    verify(watcher2).onEndpointChanged(endpointUpdateCaptor2.capture());
+    EndpointUpdate endpointUpdate2 = endpointUpdateCaptor2.getValue();
+    assertThat(endpointUpdate2.getClusterName()).isEqualTo("cluster-foo.googleapis.com");
+    assertThat(endpointUpdate2.getLocalityLbEndpointsMap())
+        .containsExactly(
+            new Locality("region1", "zone1", "subzone1"),
+            new LocalityLbEndpoints(
+                ImmutableList.of(
+                    new LbEndpoint("192.168.0.1", 8080, 2, true),
+                    new LbEndpoint("192.132.53.5", 80,5, false)),
+                1, 0));
+
+    verifyNoMoreInteractions(requestObserver);
+  }
+
+  /**
+   * An endpoint watcher is registered for a cluster that already has some other endpoint watchers
+   * watching on. Endpoint information received previously is in local cache and notified to
+   * the new watcher immediately.
+   */
+  @Test
+  public void watchEndpointsForClusterAlreadyBeingWatched() {
```

Comment:
This test is exactly the same as previous one.

Reply:
Oops. Looks so. 😢  Deleted one.